src/server/cluster/incoming_slot_migration.cc

```diff
@@ -176,13 +189,31 @@ void IncomingSlotMigration::Stop() {
       VLOG(1) << "Error during flow Stop: " << err;
     }
   }
-  bc_->Cancel();
+
+  // we need to Join the migration process to prevent data corruption
+  const absl::Time start = absl::Now();
+  const absl::Duration timeout =
+      absl::Milliseconds(absl::GetFlag(FLAGS_slot_migration_connection_timeout_ms));
+
+  while (true) {
+    const absl::Time now = absl::Now();
+    const absl::Duration passed = now - start;
+    VLOG_EVERY_N(1, 1000) << "Checking whether to continue with stop " << passed << " vs "
+                          << timeout;
+
+    if (bc_->WaitFor(absl::ToInt64Milliseconds(timeout - passed) * 1ms)) {
+      return;
+    } else if (passed >= timeout) {
+      LOG(ERROR) << "Can't stop migration in time";
+      return;
+    }
+  }
```

Comment:
This is a pretty big copy-paste, maybe we can reuse some of the code?

Reply:
I thought about it, but I don't see a good way to do it